common/cgroup/cgroup.go

```diff
@@ -69,32 +76,68 @@ func MaybeAddIsolator(opts []*unit.UnitOption, isolator string, limit *resource.
 	if limit == nil {
 		return opts, nil
 	}
-	if IsIsolatorSupported(isolator) {
+	isSupported, err := IsIsolatorSupported(isolator)
+	if err != nil {
+		return nil, err
+	}
+
+	if isSupported {
 		opts, err = isolatorFuncs[isolator](opts, limit)
-		if err != nil {
-			return nil, err
-		}
 	} else {
 		fmt.Fprintf(os.Stderr, "warning: resource/%s isolator set but support disabled in the kernel, skipping\n", isolator)
 	}
+
+	if err != nil {
+		return nil, err
+	}
 	return opts, nil
 }
 
 // IsIsolatorSupported returns whether an isolator is supported in the kernel
-func IsIsolatorSupported(isolator string) bool {
-	if files, ok := cgroupControllerRWFiles[isolator]; ok {
+func IsIsolatorSupported(isolator string) (bool, error) {
+	isUnified, err := IsCgroupUnified("/")
+	if err != nil {
+		return false, errwrap.Wrap(errors.New("error determining cgroup version"), err)
+	}
+
+	if isUnified {
+		controllers, err := GetEnabledControllers()
+		if err != nil {
+			return false, errwrap.Wrap(errors.New("error determining enabled controllers"), err)
+		}
+		for _, c := range controllers {
+			if c == isolator {
+				return true, nil
+			}
+		}
+		return false, nil
+	}
+
+	if files, ok := legacyCgroupControllerRWFiles[isolator]; ok {
 		for _, f := range files {
 			isolatorPath := filepath.Join("/sys/fs/cgroup/", isolator, f)
 			if _, err := os.Stat(isolatorPath); os.IsNotExist(err) {
-				return false
+				return false, nil
 			}
 		}
-		return true
+		return true, nil
+	}
+	return false, nil
+}
+
+// IsCgroupUnified checks if cgroup mounted at /sys/fs/cgroup is
+// the new unified version (cgroup v2)
+func IsCgroupUnified(root string) (bool, error) {
```

Comment:
I'm wondering whether it would make more sense to turn this into a more generic function which returns supported cgroup version? Result would be a `none`/`v1`/`v2` constant that could be then case-switched by consumers. Thoughts?